subiquity/server/controllers/install.py

```diff
@@ -336,6 +350,14 @@ async def run_curtin_step(name, stages, step_config, source=None):
                 for pkg in self.model.oem.metapkgs:
                     await self.install_package(package=pkg.name)
 
+            # If we already have the needed kernel installed, don't bother
+            # requesting curthooks to install it or we might end up with two
+            # kernels.
+            for pkg in await list_installed_kernels(Path(self.tpath())):
+                if pkg == self.model.kernel.needed_kernel:
```

Comment:
This is all a bit confusing but I don't think we should check this, for two reasons. One is that pkg here will be something like "linux-image-6.2.0-23-generic" or "linux-image-6.1.0-1012-oem" and I think needed_kernel is going to be a metapackage name like "linux-oem-22.04" or "linux" and the other is that even if we think we want "linux-oem-22.04" installing the metapackage might actually install "linux-oem-22.04d" or something. 

Basically IMO we should instruct curtin to not install a kernel iff there is any kernel installed already. Luckily this makes the code simpler :-)

Reply:
I'm not sure how things should interact if the autoinstall explicitly requests a different flavour though.

Reply:
Yeah, I think you're right. Dan was suggesting a runtime error if there's an explicit kernel flavor + an OEM meta-package. I think it is a sensible approach.